app/controllers/projects/new.js

```diff
@@ -0,0 +1,119 @@
+import Controller from '@ember/controller';
+import { get, getProperties, observer, set, setProperties } from '@ember/object';
+import { alias, lte } from '@ember/object/computed';
+import { inject as service } from '@ember/service';
+import { isNonValidationError } from 'code-corps-ember/utils/error-utils';
+import recordsList from 'code-corps-ember/utils/records-list';
+
+export default Controller.extend({
+  flashMessages: service(),
+  loadingBar: service(),
+  projectSkillsList: service(),
+  store: service(),
+
+  project: null,
+  showCallout: true,
+
+  isFirstProject: lte('organizationProjectsCount', 1),
+  organization: alias('project.organization'),
+  organizationProjectsCount: alias('organization.projects.length'),
+  projectCategories: alias('project.projectCategories'),
+  projectSkills: alias('project.projectSkills'),
+
+  projectTitleChanged: observer('project.title', function() {
+    let title = get(this, 'project.title');
+    let slug = title ? title.dasherize() : '';
+    set(this, 'project.slug', slug);
+  }),
+
+  addCategory(category) {
+    let { project, store } = getProperties(this, 'project', 'store');
+    let projectCategory = store.createRecord('project-category', { project, category });
+    project.get('projectCategories').pushObject(projectCategory);
+  },
+
+  removeCategory(category) {
+    let { project, projectCategories }
+      = getProperties(this, 'project', 'projectCategories');
+
+    let projectCategory = recordsList.find(projectCategories, category, project);
+    projectCategory.destroyRecord();
+  },
+
+  removeProjectSkill(projectSkill) {
+    return projectSkill.destroyRecord();
+  },
+
+  toggleSkill(skill) {
+    let list = get(this, 'projectSkillsList');
+    return list.toggleWithoutSave(skill);
+  },
+
+  updateCategories(newSelection, value, operation) {
+    if (operation === 'added') {
+      this.addCategory(value);
+    }
+
+    if (operation === 'removed') {
+      this.removeCategory(value);
+    }
+  },
+
+  uploadDone(cloudinaryPublicId) {
+    let project = get(this, 'project');
+    set(project, 'cloudinaryPublicId', cloudinaryPublicId);
+    this._stopLoadingBar();
+  },
+
+  uploadErrored() {
+    this._stopLoadingBar();
+    get(this, 'flashMessages').clearMessages().danger('Upload failed');
+  },
+
+  uploadStarted() {
+    this._startLoadingBar();
+  },
+
+  actions: {
+    copyOrganizationData(organization) {
+      let { cloudinaryPublicId, description, iconLargeUrl, name, slug }
+        = getProperties(organization, 'cloudinaryPublicId', 'description', 'iconLargeUrl', 'name', 'slug');
+      let project = get(this, 'project');
+      let attrs = { cloudinaryPublicId, description, iconLargeUrl, slug, title: name };
+      setProperties(project, attrs);
+      set(this, 'showCallout', false);
+      get(this, 'flashMessages').clearMessages().success('Your project now has the details from your organization.');
+    },
```

Comment:
This whole thing feels complex enough to warrant extracting the form out into a component. Thoughts here? It will get even worse with client-side validations.

We could also try to reduce some of the acceptance tests with more usage of route/controller unit tests.

Reply:
If you feel it's worth the potential speed gain of integration tests instead of acceptance tests to invest the time right now, I'm ok with it. 

I'm hesitant to abandon the base acceptance tests in favor of faster integration tests. Acceptance tests are better at ensuring the overall app behavior works and we're risking not detecting some layer not working as expected when it interfaces with another layer.

As an example, I think we need an acceptance test for validation errors, but we could use integration tests to ensure errors on all supported fields are rendered.

For navigation confirmation tests alone, it's difficult to do integration tests well, without mocking large parts of the app, which sort of defeats the purpose.

Reply:
Not saying we abandon the acceptance tests, but all the edge cases would be much simpler to test with integration tests.